File: src/jl/server_interface.jl


module ServerInterface
    abstract type MethodParams end

    import JSONRPC

    # JSONRPC.@dict_readable macro creates StructName(d::Dict) abd StructName(;field::type,...) constructors
    # make automatic conversion possible by calling this constructor
    Base.convert(::Type{T}, d::Dict) where T <: JSONRPC.Outbound = T(d)
    Base.convert(::Type{T}, d::Dict) where T <: MethodParams = T(d)

    # request and response types

    JSONRPC.@dict_readable struct SyntaxNode <: JSONRPC.Outbound
        node_id::String
        first_byte::Int
        last_byte::Int
        source_text::String
    end

    JSONRPC.@dict_readable struct ControlNode <: JSONRPC.Outbound
        node::SyntaxNode
        kind::String
        control_subnode::SyntaxNode
        body::Vector{SyntaxNode}
    end

    JSONRPC.@dict_readable struct CallGraphNode <: JSONRPC.Outbound
        caller::SyntaxNode
        called::Vector{SyntaxNode}
    end


    JSONRPC.@dict_readable struct Model <: JSONRPC.Outbound
        name::String
        node::SyntaxNode
    end

    JSONRPC.@dict_readable struct DistributionParam <: JSONRPC.Outbound
        name::String
        node::SyntaxNode
    end

    JSONRPC.@dict_readable struct Distribution <: JSONRPC.Outbound
        name::String
        node::SyntaxNode
        params::Vector{DistributionParam}
    end

    JSONRPC.@dict_readable struct RandomVariable <: JSONRPC.Outbound
        node::SyntaxNode
        name::String
        distribution::Distribution
        is_observed::Bool
    end

    JSONRPC.@dict_readable struct File <: JSONRPC.Outbound
        file_name::String
        ppl::String
    end

    JSONRPC.@dict_readable struct Interval <: JSONRPC.Outbound
        low::String # to allow inf
        high::String
    end

    JSONRPC.@dict_readable struct SymbolicExpression <: JSONRPC.Outbound
        expr::String
    end

    # endpoints

    const build_ast_rt = JSONRPC.RequestType("build_ast", File, String)

    const get_model_rt = JSONRPC.RequestType("get_model", File, Model)
    
    const get_guide_rt = JSONRPC.RequestType("get_guide", File, Model)

    const get_random_variables_rt = JSONRPC.RequestType("get_random_variables", File, Vector{RandomVariable})

    JSONRPC.@dict_readable struct tree_node_params <: MethodParams
        tree_id::String
        node::SyntaxNode
    end

    const get_data_dependencies_rt = JSONRPC.RequestType("get_data_dependencies", tree_node_params, Vector{SyntaxNode})

    const get_control_parents_rt = JSONRPC.RequestType("get_control_parents", tree_node_params, Vector{ControlNode})

    JSONRPC.@dict_readable struct estimate_value_range_p <: MethodParams
        tree_id::String
        expr::SyntaxNode
        assumptions::Vector{Tuple{RandomVariable,Interval}}
    end
    const estimate_value_range_rt = JSONRPC.RequestType("estimate_value_range", estimate_value_range_p, Interval)

    const get_call_graph_rt = JSONRPC.RequestType("get_call_graph", tree_node_params, Vector{CallGraphNode})

    JSONRPC.@dict_readable struct get_path_conditions_p <: MethodParams
        tree_id::String
        root::SyntaxNode
        nodes::Vector{SyntaxNode}
        assumptions::Vector{Tuple{SyntaxNode,SymbolicExpression}}
    end
    const get_path_conditions_rt = JSONRPC.RequestType("get_path_conditions", get_path_conditions_p, Vector{SymbolicExpression})

end

File: src/static/lasapp/server_interface.py
from dataclasses import dataclass
from dataclasses_json import dataclass_json

@dataclass_json
@dataclass
class SyntaxNode:
    node_id: str
    first_byte: int
    last_byte: int
    source_text: str

    def __hash__(self) -> int:
        return self.node_id.__hash__()
    def __eq__(self, other) -> bool:
        if isinstance(other, SyntaxNode):
            return self.node_id == other.node_id
        return False

@dataclass_json 
@dataclass
class ControlDependency:
    node: SyntaxNode
    kind: str
    control_node: SyntaxNode
    body: list[SyntaxNode]

@dataclass_json 
@dataclass
class CallGraphNode:
    caller: SyntaxNode
    called: list[SyntaxNode]


@dataclass_json
@dataclass
class Model:
    name: str
    node: SyntaxNode

@dataclass_json
@dataclass
class DistributionParam:
    name: str
    node: SyntaxNode

@dataclass_json
@dataclass
class Distribution:
    name: str
    node: SyntaxNode
    params: list[DistributionParam]

@dataclass_json
@dataclass
class RandomVariable:
    node: SyntaxNode
    name: str
    address_node: SyntaxNode
    distribution: Distribution
    is_observed: bool
    def __eq__(self, __value: object) -> bool:
        if isinstance(__value, RandomVariable):
            return self.node.node_id == __value.node.node_id
        return False
    def __hash__(self) -> int:
        return hash(self.node.node_id)
    def __repr__(self) -> str:
        return f"RandomVariable({self.node.source_text})"


@dataclass_json
@dataclass
class Interval:
    low: str
    high: str


@dataclass_json
@dataclass
class SymbolicExpression:
    expr: str
